### ISSUE TITLE
No version of Visual Studio compiler found with SCons 3.0.2+ and VS 2017

### ISSUE THREAD

**Issue by girng** [godotengine/godot#24809]
**Godot version:**
3873f84548cd91ec62b9cb045e75158df74e19f9

**OS/device including version:**
w 10, 64-bit

**Issue description:**
- when trying to compile godot, this error displays:
```
 warning: No version of Visual Studio compiler found - C/C++ compilers most likely not set correctly
```
- what i found odd is,  it says no compiler found, but still compiles. 
- and what's even more odd, when you edit a file and recompile, it recompiles everything. which isn't normal.. because before, it would just recompile that 1 file you modified.  🤔🤷‍♀️🤦‍♂️

**Steps to reproduce:**
- download 2017 build tools here https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2017 (make sure you have Windows SDK enabled)
- install python 3.7.2
- install scons v3.0.2

i'm running the scons command when i have the build tools cmd open, it's displaying:
```
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.0
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'

C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools>
```

**Comment by akien-mga** [comment]
Did you install the C++ compilers? https://docs.godotengine.org/en/3.0/development/compiling/compiling_for_windows.html#installing-visual-studio-caveats

**Comment by girng** [comment]
> Did you install the C++ compilers? https://docs.godotengine.org/en/3.0/development/compiling/compiling_for_windows.html#installing-visual-studio-caveats

yep! they were checked in the vscode installer. i am not sure how to debug this, should i start with changing the scons version, downgrading to 2015 tools, or?

also, here's the exact info that points to a line in the SContruct file
```
scons: warning: No version of Visual Studio compiler found - C/C++ compilers most likely not set correctly
File "C:\Godot Builds\godot-master\SConstruct", line 3, in <module>

scons: warning: No version of Visual Studio compiler found - C/C++ compilers most likely not set correctly
File "C:\Godot Builds\godot-master\SConstruct", line 63, in <module>
Configuring for Windows: target=release_debug, bits=default
Found MSVC compiler: amd64
Compiled program architecture will be a 64 bit executable (forcing bits=64).
```

line 3 is `EnsureSConsVersion(0, 98, 1)`
and line 63 is:
`env_base = Environment(tools=custom_tools)`

**Comment by girng** [comment]
thank you @ByTheQuietLake for your testing.

i just did some more testing as well. i tried scons **3.0.4** (latest release). and unfortunately it still re-compiles every file instead of just the one you modified (even after you built godot).
___
on windows, you can use `pip install scons==3.0.1` 👍 
if you get this error after running that command:

```
error: option --single-version-externally-managed not recognized
```

 please follow [this answer](https://stackoverflow.com/a/15258644/7873241) and run:

```
pip install -U setuptools
pip install -U wheel
```
now run `pip install scons==3.0.1`

okay. seems like even with scons 3.0.1, it still re-compiles every file.. what the heck. going to try python 2.7

**Comment by CarysT** [comment]
Just to add to this I have Scons version 3.0.5 installed, and VS2019. I'm on Windows10 64bit, and when I run scons to build it, It gives me the same "error" but still proceeds to compile the files because it does find a MSVC compiler. I have no idea why it would be complaining then though.

```**********************************************************************
** Visual Studio 2019 Developer Command Prompt v16.1.0-pre.1.0
** Copyright (c) 2019 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'

C:\godot>scons -j12 platform=windows
scons: Reading SConscript files ...

scons: warning: No version of Visual Studio compiler found - C/C++ compilers most likely not set correctly
File "C:\godot\SConstruct", line 3, in <module>

scons: warning: No version of Visual Studio compiler found - C/C++ compilers most likely not set correctly
File "C:\godot\SConstruct", line 64, in <module>
Configuring for Windows: target=debug, bits=default
Found MSVC compiler: amd64

**Comment by PtrMan** [comment]
I hot the same bug with funny variations

Running from the MS VS 2019 Community pre release

```
**********************************************************************
** Visual Studio 2019 Developer Command Prompt v16.1.0-pre.1.0
** Copyright (c) 2019 Microsoft Corporation
**********************************************************************

S:\win10host\programs\dev\ide\visual studio express 2019 preview>cd C:\godot

...

C:\godot>scons p=windows
scons: Reading SConscript files ...

scons: warning: No version of Visual Studio compiler found - C/C++ compilers most likely not set correctly
File "C:\godot\SConstruct", line 3, in <module>

scons: warning: No version of Visual Studio compiler found - C/C++ compilers most likely not set correctly
File "C:\godot\SConstruct", line 64, in <module>
Configuring for Windows: target=debug, bits=default
Using Mingw
WindowsError: [Error 2] Das System kann die angegebene Datei nicht finden:
  File "C:\godot\SConstruct", line 317:
    version = methods.get_compiler_version(env)
  File "C:\godot\methods.py", line 605:
    version = decode_utf8(subprocess.check_output([env['CXX'], '--version']).strip())
  File "C:\python27-x64\lib\subprocess.py", line 212:
    process = Popen(stdout=PIPE, *popenargs, **kwargs)
  File "C:\python27-x64\lib\subprocess.py", line 390:
    errread, errwrite)
  File "C:\python27-x64\lib\subprocess.py", line 640:
    startupinfo)
```
(it somehow reasons that it's best to compile with mingw even if it doesn't exist)

And running from a 2017 prompt

```
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.9.11
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x86'

C:\Program Files (x86)\Microsoft Visual Studio\2017\Community>C:\godot
Der Befehl "C:\godot" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.

C:\Program Files (x86)\Microsoft Visual Studio\2017\Community>C.
Der Befehl "C." ist entweder falsch geschrieben oder
konnte nicht gefunden werden.

C:\Program Files (x86)\Microsoft Visual Studio\2017\Community>C:

C:\Program Files (x86)\Microsoft Visual Studio\2017\Community>cd C:\godot

C:\godot>scons p=windows vsproj=yes
scons: Reading SConscript files ...

scons: warning: No version of Visual Studio compiler found - C/C++ compilers most likely not set correctly
File "C:\godot\SConstruct", line 3, in <module>

scons: warning: No version of Visual Studio compiler found - C/C++ compilers most likely not set correctly
File "C:\godot\SConstruct", line 64, in <module>
Configuring for Windows: target=debug, bits=default
Found MSVC compiler: x86
Compiled program architecture will be a 32 bit executable. (forcing bits=32).
KeyError: 'MSVC_VERSION':
  File "C:\godot\SConstruct", line 294:
    detect.configure(env)
  File "./platform/windows\detect.py", line 369:
    configure_msvc(env, manual_msvc_config)
  File "./platform/windows\detect.py", line 199:
    if int(env['MSVC_VERSION'].split('.')[0]) >= 14: #vs2015 and later
  File "C:\python27-x64\scons\SCons\Environment.py", line 410:
    return self._dict[key]

C:\godot>
```

maybe it has an issue that the compiler is installed on another drive, no idea.
It's just another bugged stupid tool.

**Comment by PtrMan** [comment]
I solved it by installing the windows 10 SDK from the Microsoft Visual Studio 2019 installer and by updating scons to the newest version, see https://godotengine.org/qa/44344/unable-to-build-godot-from-source-on-windows for details

**Comment by Anutrix** [comment]
For VS2017, the following must be installed:

- Desktop development with C++
- - From Optional(VC++ 2017 version 15.9 v14.16 latest v141 tools). ; Mandatory for compiling Godot.

**Comment by girng** [comment]
I solved this issue by downloading scons 3.1.0 here https://scons.org/pages/download.html. Extracted the zip, and executed `python setup.py install`
